client/lib/media/actions.js

```diff
@@ -116,13 +117,18 @@ MediaActions.add = function( siteId, files ) {
 			date: new Date( baseTime - ( files.length - i ) ).toISOString()
 		};
 
-		if ( 'string' === typeof file ) {
-			// Generate from string
+		// Determine upload mechanism by object type
+		const isUrl = 'string' === typeof file;
+		const addHandler = isUrl ? 'addMediaUrls' : 'addMediaFiles';
+
+		if ( isUrl ) {
+			// Generate from URL
+			const filePath = url.parse( file ).pathname;
```

Comment:
I feel as though we should enhance the existing `MediaUtils` helpers to pull the correct extension from the path name, since other usages will still be susceptible to the issues we've experienced here (it's also easier to add [test cases](https://github.com/Automattic/wp-calypso/blob/master/client/lib/media/test/utils.js)).

In the course of doing so, we may need to add a new method `MediaUtils.getFileExtension`.